src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -210,29 +229,59 @@ namespace AppInstaller::CLI::Workflow
             return;
         }
 
-        // Get a mapping of source IDs to names so that we can show something nicer
-        std::map<std::string, std::string> sourceNames;
-        for (const auto& source : Repository::Source::GetCurrentSources())
-        {
-            sourceNames[source.Identifier] = source.Name;
-        }
-
-        Execution::TableOutput<4> table(context.Reporter,
+        Execution::TableOutput<6> table(context.Reporter,
             {
+                Resource::String::SearchName,
                 Resource::String::SearchId,
+                Resource::String::SearchVersion,
                 Resource::String::SearchSource,
                 Resource::String::PinType,
-                Resource::String::SearchVersion,
+                Resource::String::PinVersion,
             });
 
+        const auto& source = context.Get<Execution::Data::Source>();
         for (const auto& pin : pins)
         {
-            table.OutputLine({
-                pin.GetPackageId(),
-                sourceNames[pin.GetSourceId()],
-                std::string{ ToString(pin.GetType()) },
-                pin.GetGatedVersion().ToString(),
+            const auto& pinKey = pin.GetKey();
+            auto searchRequest = GetSearchRequestForPin(pin.GetKey());
+            auto searchResult = source.Search(searchRequest);
+            for (const auto& match : searchResult.Matches)
+            {
+                Utility::LocIndString packageName;
+                Utility::LocIndString sourceName;
+                Utility::LocIndString version;
+
+                if (pinKey.IsForInstalled())
+                {
+                    sourceName = Resource::LocString{ Resource::String::PinInstalledSource };
+                }
+                else
+                {
+                    // This ensures we get the info from the right source if it exists on multiple
+                    auto availableVersion = match.Package->GetAvailableVersion({ pinKey.SourceId, "", "" });
+                    if (availableVersion)
+                    {
+                        packageName = availableVersion->GetProperty(PackageVersionProperty::Name);
+                        sourceName = availableVersion->GetProperty(PackageVersionProperty::SourceName);
+                    }
+                }
+
+                auto installedVersion = match.Package->GetInstalledVersion();
+                if (installedVersion)
+                {
+                    packageName = installedVersion->GetProperty(PackageVersionProperty::Name);

```

Comment:
I think if we have it installed it's better to show that name instead of the available. That way it matches with what we show in list/upgrade commands. It also makes the differences clearer if a single available pin matches multiple installed packages.